From 6892c5513bb29d697bd20faceafcaed9ed2d080c Mon Sep 17 00:00:00 2001
From: Chad Dupuis <cdupuis@redhat.com>
Date: Thu, 21 Aug 2014 13:06:42 -0400
Subject: [ethernet] qlcnic: Fix panic while dumping TX queues on TX timeout

Message-id: <1408626422-3941-119-git-send-email-cdupuis@redhat.com>
Patchwork-id: 88631
O-Subject: [PATCH RHEL 7.1 118/138] qlcnic: Fix panic while dumping TX queues on TX timeout
Bugzilla: 1089352
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Manish Chopra <manish.chopra@qlogic.com>

Bug 1089352 (https://bugzilla.redhat.com/show_bug.cgi?id=1089352)

net commit id d5d2bf3eabb34cc8eaf54db37fdc43f04267985a

o In case of non-multi TX queue mode driver does not initialize "crb_intr_mask" pointer
  and driver was accessing that un-initialized pointer while dumping TX queue.
  So dump "crb_intr_mask" only when it is initilaized.

Signed-off-by: Manish Chopra <manish.chopra@qlogic.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/qlogic/qlcnic/qlcnic_main.c b/drivers/net/ethernet/qlogic/qlcnic/qlcnic_main.c
index 2ddd0f9..0dc8615 100644
--- a/drivers/net/ethernet/qlogic/qlcnic/qlcnic_main.c
+++ b/drivers/net/ethernet/qlogic/qlcnic/qlcnic_main.c
@@ -2959,9 +2959,13 @@ static void qlcnic_dump_tx_rings(struct qlcnic_adapter *adapter)
        tx_ring->tx_stats.xmit_called,
        tx_ring->tx_stats.xmit_on,
        tx_ring->tx_stats.xmit_off);
+
+  if (tx_ring->crb_intr_mask)
+   netdev_info(netdev, "crb_intr_mask=%d\n",
+        readl(tx_ring->crb_intr_mask));
+
   netdev_info(netdev,
-       "crb_intr_mask=%d, hw_producer=%d, sw_producer=%d sw_consumer=%d, hw_consumer=%d\n",
-       readl(tx_ring->crb_intr_mask),
+       "hw_producer=%d, sw_producer=%d sw_consumer=%d, hw_consumer=%d\n",
        readl(tx_ring->crb_cmd_producer),
        tx_ring->producer, tx_ring->sw_consumer,
        le32_to_cpu(*(tx_ring->hw_consumer)));
-- 
1.7.1